File: cpp/p15.cpp
#include "euler.h"

using namespace std;

vector<int> get_nth_row_in_pascal_triangle( int n ) {
  int r = n + 1, v = 1;
  vector<int> output( 1, v );
	
  for( int c = 1; c < r; c++ ) {
    v = v * ( r - c ) / c;
    output.push_back( v );
  }
	
  return output;
}

int main() {
  int size = 20;
  long long num_paths = 0;

  vector<int> row = get_nth_row_in_pascal_triangle( size );
	
  for( int i = 0; i < row.size(); i++ )
    num_paths += (long long) row [ i ] * row [ i ];
		
  cout << num_paths << endl;
}

File: clojure/p15.clj
(ns p15
  (:use [euler :only (n-choose-k timed-test)]))

;; # paths in NxN grid = sum of squares of Nth row of Pascal's triangle
;; sum of squares of Nth row of Pascal's triangle = (2N choose N)
;; (http://en.wikipedia.org/wiki/Pascal_triangle#Rows)

(timed-test
 137846528820
 (let [N 20]
   (n-choose-k (* 2 N) N)))
